[Dynamic Dashboard] Extract stats card

hichamboushaba opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of: #11219 

### Description
This PR is a first step to implement #11219, it extracts the stats logic to:
1. A separate ViewModel.
2. The UI to a Composable function.

Sorry for the large number of changes, I didn't expect it to be this big, but if it's an issue, just let me know, and I'll try to split it, I'll try also to put some comments in the code to help with the review.
The first commit 0b34c234a9050235f464c8f9badc337531a2f16d can be reviewed separately, as it's just bringing changes of https://github.com/woocommerce/woocommerce-android/pull/11191 to the new package `dashboard`.

### Testing instructions
Smoke test the stats in the MyStore screen, and confirm it works as expected.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->


- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

hichamboushaba committed:
Bring changes related to the summary visitor stats

See https://github.com/woocommerce/woocommerce-android/pull/11191

hichamboushaba committed:
Extract logic of observing the selected range to an external usecase

hichamboushaba committed:
First pass at extracting the stats logic to a separate viewmodel

Also it includes moving the view to a Composable function

hichamboushaba committed:
Move events handling to the card

hichamboushaba committed:
Fix detekt issues

hichamboushaba committed:
Handle visitor stats

dangermattic commented:
<!--
  0 Errors
  5 Warnings: View files have been modified,...
  1 Message: <p>This PR contains changes to...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          5 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>DashboardStatsViewModel</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Class <code>GetSelectedDateRange</code> is missing tests, but <code>unit-tests-exemption</code> label was set to ignore this.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is not assigned to a milestone.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false"><p>This PR contains changes to Tracks-related logic. Please ensure (<strong>author and reviewer</strong>) the following are completed:</p>

<ul>
  <li>The tracks events must be validated in the Tracks system.</li>
  <li>Verify the internal Tracks spreadsheet has also been updated.</li>
  <li>Please consider registering any new events.</li>
  <li>The PR must be assigned the <strong>category: tracks</strong> label.</li>
</ul></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hichamboushaba committed:
Use all the properties as keys for the LaunchedEffect

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/dashboard/DashboardFragment.kt:
```diff
@@ -177,23 +150,33 @@ class DashboardFragment :
             storeOnboardingViewModel.onPullToRefresh()
-            binding.myStoreStats.clearStatsHeaderValues()
-            binding.myStoreStats.clearChartData()
+//            binding.myStoreStats.clearStatsHeaderValues()
+//            binding.myStoreStats.clearChartData()
```
I'll try to handle pull-to-refresh separately, preferably when we have a decision p1712147928963089-slack-C03L1NF1EA3, and if it takes time, I'll have something simple in a separate PR.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/dashboard/DashboardFragment.kt:
```diff
@@ -177,23 +150,33 @@ class DashboardFragment :
+                    currencyFormatter = currencyFormatter,
+                    usageTracksEventEmitter = usageTracksEventEmitter,
+                    onPluginUnavailableError = { updateStatsAvailabilityError() },
+                    reportJetpackPluginStatus = { onVisitorStatsUnavailable(it) },
```
I might probably move the Jetpack banner logic to the parent fragment in a separate PR, it seems weird to handle this in a child card.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/dashboard/stats/GetSelectedDateRange.kt:
```diff
@@ -0,0 +1,61 @@
+import java.util.Locale
+import javax.inject.Inject
+
+class GetSelectedDateRange @Inject constructor(
```
This usecase helps us with the syncing of the selected date between the stats and the top performers card.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/dashboard/stats/DashboardStatsViewModel.kt:
```diff
@@ -0,0 +1,336 @@
+        super.onCleared()
+    }
+
+    private suspend fun loadStoreStats(selectedRange: StatsTimeRangeSelection, forceRefresh: Boolean) = coroutineScope {
```
Without the `coroutineScope` here, the `launch` call at line `192` was leaked, and we were observing the `lastUpdate` multiple times.

hichamboushaba committed:
Use a coroutineScope builder to avoid leaking a job

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/dashboard/DashboardViewModel.kt:
```diff
@@ -167,29 +168,30 @@ class DashboardViewModel @Inject constructor(
-                    mapOf(AnalyticsTracker.KEY_RANGE to selectedRange.selectionType.identifier)
-                )
+    private suspend fun loadTopPerformersStats(selectedRange: StatsTimeRangeSelection, forceRefresh: Boolean) =
+        coroutineScope {
```
Same as https://github.com/woocommerce/woocommerce-android/pull/11231#discussion_r1551765361

hichamboushaba committed:
Move GetStats to the correct package

JorgeMucientes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/dashboard/DashboardFragment.kt:
```diff
@@ -177,23 +150,33 @@ class DashboardFragment :
+        binding.myStoreStats.apply {
+            setViewCompositionStrategy(ViewCompositionStrategy.DisposeOnViewTreeLifecycleDestroyed)
+
+            setContent {
+                DashboardStatsCard(
```
~~I think we are missing `WooThemeWithBackground {` here or else that card won't look as expected in dark mode:~~ 

I haven't dig deeper into why this happens in dark mode, but adding `WooThemeWithBackground` won't fix the issue as I though it would. 


https://github.com/woocommerce/woocommerce-android/assets/2663464/c6434c49-ccaf-46c4-baf9-d6af9d71217a

hichamboushaba replied:
The issue is that we weren't setting the correct style to the view, as it was hardcoded in the XML layout, but since now the view is just a ComposeView, it didn't work.
I updated the view in https://github.com/woocommerce/woocommerce-android/pull/11231/commits/d2ede6e67cef808127c2451c59655880d1aae416 to use the correct style attribute, which we correctly set in our theme, so this fixes the issue.

JorgeMucientes commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/dashboard/DashboardStatsView.kt:
```diff
@@ -606,6 +609,14 @@ class DashboardStatsView @JvmOverloads constructor(
     }
 
+    private fun hideVisitorStats() {
+        binding.statsViewRow.visitorsValueTextview.isVisible = false
+        binding.statsViewRow.emptyVisitorStatsIndicator.isVisible = true
+        binding.statsViewRow.emptyVisitorStatsIcon.isVisible = true
+        binding.statsViewRow.conversionValueTextView.isVisible = false
+        binding.statsViewRow.emptyConversionRateIndicator.isVisible = true
```
I think we are missing the `binding.statsViewRow.emptyVisitorStatsIndicator.isVisible = true` here to show the thick dash when hiding visitor numbers.

hichamboushaba replied:
It's on line 614, what issue was this about?
I found one issue about wrongly showing the visitor count on the custom tab sometimes, and I'm working on a solution, so if this was just about this, I'll fix it on a separate PR.

hichamboushaba replied:
@JorgeMucientes I ended making a structural change to how we update the view in 28c9455203a4d42d89133817f1bd9a1c71bc8cee, this fixes the issue that I noticed, let me know if you had a different issue.

In the commit, I added a comment explaining why the change, please let me know if it's not clear enough.

JorgeMucientes replied:
Hey @hichamboushaba the issue I noticed was that we do not display the `dash` aka `emptyVisitorStatsIndicator` when failing to load visitor stats. 

https://github.com/woocommerce/woocommerce-android/assets/2663464/1557de20-5079-4707-872b-c3c4d318439a

This is probably unrelated to the changes from this PR and is related to the prior PR where we discussed this already. But since this PR applied some refactoring to the view I though it could be a good chance to fix it. However, my suggestion was not correct. The problem lies with `binding.statsViewRow.emptyVisitorsStatsGroup.isVisible = false`. Whenever we set the group to false it doesn't matter if we add `binding.statsViewRow.emptyVisitorStatsIndicator.isVisible = true`, it won't have any impact. 
To properly fix this there are several places were we need to change how to handle this, or maybe we can even get rid of `emptyVisitorsStatsGroup`. 
In any case it feels out of scope for the PR. So we can address it elsewhere.

hichamboushaba replied:
I see, as I didn't touch anything in the view itself in this PR, I misunderstood the issue.

> But since this PR applied some refactoring to the view I though it could be a good chance to fix it.

This is the origin of the confusion, I didn't do any refactoring to the legacy view, all the changes to the `DashboardStatsView` were just about bringing the changes of the last PR #11191.

We can address the issue in a separate PR.

JorgeMucientes commented:
Hey @hichamboushaba, excellent work 🥇. Really neat refactor!
I just noticed a couple of issues that it would be nice to address before merging.

hichamboushaba committed:
Fix style of the dashboard stats view

hichamboushaba committed:
Move stats view to correct package

hichamboushaba committed:
Make function private

hichamboushaba committed:
Merge branch 'trunk' into issue/11219-extract-stats-card

# Conflicts:
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/dashboard/DashboardFragment.kt
#	WooCommerce/src/main/kotlin/com/woocommerce/android/ui/dashboard/DashboardViewModel.kt

hichamboushaba committed:
Update DashboardStatsCard for a better handling of updates

The update uses Side effects for granular updates instead of using the built in update function of AndroidView.

hichamboushaba committed:
Fix detekt issues
